Added retry logic to minaTransactionSender

Raunaque97 opened the pull request:
closes #340 

- refactored minaTransactionSender to use a persistent Storage.
- added a simple retry Strategy to increase fees by a multiple and retry every few minutes if the transaction is not included.
- also handles the case if the sequencer restarts and the last sent transaction statuses were not updated in DB.

Raunaque97 committed:
added PendingL1TransactionStorage for persisting mina transactions in db.

Raunaque97 committed:
implementing PendingL1TransactionStorage

Raunaque97 committed:
implementing L1TransactionRetryStrategy

Raunaque97 committed:
added MinaSigner interface

Raunaque97 committed:
updated MinaTransactionSender to use L1TransactionRetryStrategy and PendingL1TransactionStorage

Raunaque97 committed:
fixing tests

Raunaque97 committed:
refactoring, removed feeStrategy dependencies from bridgingModule, settlementModule

rpanic commented on packages/persistance/prisma/schema.prisma:
```diff
@@ -148,3 +148,15 @@ model IncomingMessageBatch {
   messages IncomingMessageBatchTransaction[]
 }
+
+model PendingL1Transaction {
```
You forgot to create a migration for this new schema change - do this by cd-ing into the persistence pkg, then run `prisma migrate dev` and it'll you for a name and migrate

rpanic commented:
So what I just discovered: When we setFee(), o1js erases all previous signatures in order for the tx to be signed again. And this makes sense - for the feepayer and _some_ accountupdates, o1js uses the so-called "full commitment". This is simply just the hash of the whole transaction, including all accountupdates and the feepayer, whereas the non-full commitment would just the be hash of the current accountupdate. 
Now, accountupdates can actually specify whether they want to use the full commitment or not (this is a field in the accountupdate struct that mina will receive as part of the transaction). And whatever commitment you choose, you signature or proof will only be valid given that exact commitment. it is kinda the basis on which you sign or prove. Now why is this important?
Signing or proving over a full commitment prevent replay attacks, like in other chains. But mina is a bit special, and unlike other chains, we sign a bunch of sub-parts individually. And so generally, in order to prevent somebody to rip out a part of our tx from the L1 mempool and replay it in their own specially-crafted malicious transaction and frontrun ours, we sign over the full commitment. 
For resending transactions with a higher fee, this is a bit annoying - we don't want to re-prove the entire tx just because we changed the fee, right?
As I've mentioned above, luckily we can choose to opt-out of using the full commitment and use only the current accountupdate's commitment instead. But this is only possible in some distinct scenarios. My initial suspicious is that this is the case iff the accountupdate's validity is directly dependent on exclusively other full-commitment based accountupdates within the same transaction. But I'll have to think more on that definition.
Anyways, what we have to do before this PR will work without re-proving some accountupdates, is to use no full commitment generally. And for us to be able to do this, we need to examine the protocol and see if we can do this safely

rpanic commented on packages/sequencer/src/settlement/transactions/DefaultL1TransactionRetryStrategy.ts:
```diff
@@ -0,0 +1,90 @@
+    const currentFee = tx.transaction.feePayer.body.fee;
+    const newFee = UInt64.from(this.bumpFee(Number(currentFee.toBigInt())));
+    tx.setFee(newFee);
+    // Delay if needed
```
Can you explain what this delay here accomplishes? I'd expect this function to only craft the new transaction and return it...

Raunaque97 replied:
This is used to add a delay between retries, like retrying a transaction every 5 minutes.

rpanic commented on packages/sequencer/src/storage/repositories/PendingL1TransactionStorage.ts:
```diff
@@ -0,0 +1,34 @@
+  nonce: number;
+  attempts: number;
+  status: PendingL1TransactionStatus;
+  transactionJson: string;
```
Couldn't this struct here hold the transaction directly? This way we don't have to parse it at all for in-memory mode, and then the Mapper can handle parsing for the Prisma variant

rpanic commented on packages/persistance/prisma/schema.prisma:
```diff
@@ -148,3 +148,15 @@ model IncomingMessageBatch {
+  lastError      String?
+  sentAt         DateTime? 
+
+  @@id([sender, nonce])
```
I'm not against this schema, but I think for the reorg support we'll have to change this - but that will happen when we tackle that piece of work.

rpanic replied:
What I'd love to see would be if we could add foreign-key out of settlement here. So in the Settlement entity, add a reference to the PendingTransaction so that we can establish that link

rpanic commented on packages/persistance/src/services/prisma/PrismaPendingL1TransactionStorage.ts:
```diff
@@ -0,0 +1,103 @@
+        },
+      },
+      data: {
+        ...(updates.attempts !== undefined && { attempts: updates.attempts }),
```
Why do you have this here? Since it's partial already, couldn't you just use the `updates` object here directly?

rpanic commented on package.json:
```diff
@@ -74,6 +74,7 @@
     ]
   },
   "dependencies": {
+    "clean": "^4.0.2",
```
This dependency isn't used afaict

rpanic commented on packages/sequencer/src/settlement/transactions/DefaultL1TransactionRetryStrategy.ts:
```diff
@@ -0,0 +1,90 @@
+    const tx = Transaction.fromJSON(JSON.parse(record.transactionJson));
+    const currentFee = tx.transaction.feePayer.body.fee;
+    const newFee = UInt64.from(this.bumpFee(Number(currentFee.toBigInt())));
+    tx.setFee(newFee);
```
As mentioned in the comment earlier, we need to figure something out to prevent re-proving

rpanic commented on packages/sequencer/src/settlement/transactions/DefaultL1TransactionRetryStrategy.ts:
```diff
@@ -0,0 +1,90 @@
+};
+
+@sequencerModule()
+export class DefaultL1TransactionRetryStrategy
```
I like this encapsulation!

Raunaque97 replied:
😀

rpanic commented on packages/sequencer/src/settlement/transactions/MinaTransactionSender.ts:
```diff
@@ -31,114 +33,69 @@ export type TxSendResult<Input extends "sent" | "included" | "none"> =
+    private readonly retryStrategy: L1TransactionRetryStrategy,
+    @inject("MinaSigner") private readonly signer: MinaSigner
+  ) {
+    void this.startPolling();
```
Lets not void this completely, instead of a while(true), lets create a interval, save the id, and then delete the interval in `close()` (which we have to add using `@closeable()` and `implements Closeable` on this class).
If we don't do this, appchain.close() won't end the process bcs there's still this dangling promise running in the background - this is especially annoying since tests won't terminate

rpanic reviewed (changes requested):
This is great! I have a few minor requests and questions, but I like it generally

Raunaque97 committed:
using SettlementSigner, and removing LocalMinaSigner

Raunaque97 committed:
storing transaction obj in PendingL1TransactionRecord instead of json string

Raunaque97 committed:
making MinaTransactionSender closeable

Raunaque97 committed:
Refactor PendingL1Transaction and Settlement models to use transactionId instead of transactionHash,
PendingL1TransactionStorage to use id as the primary identifier.

Raunaque97 committed:
added a queued wait status in MinaTransactionSender

Raunaque97 committed:
adding foreign key relation between Settlement and  PendingL1Transaction

Raunaque97 committed:
fixing lint errors

Raunaque97 committed:
created the migration file

Raunaque97 committed:
refactoring, renaming mappers

Raunaque97 committed:
lint fix

Raunaque97 committed:
Add hash field to PendingL1Transaction model and added txn status polling

Raunaque97 committed:
added  minaTransactionSender tests

Raunaque97 committed:
updating PendingL1Transaction schema to include nextActionAt and queuedAt

Raunaque97 committed:
refactoring MinaTransactionSender.
- polling the chain and updating persistant storage is now handled by L1TransactionDispatcher
- TxStatusWaiter is used for waiting on a txn to get sent, included.

Raunaque97 committed:
fixing build errors after rebase

Raunaque97 committed:
moved transactionDispatcher to  MinaBaseLayer

Raunaque97 committed:
using signProveAndSendTransaction in place of proveAndSendTransaction

Raunaque97 committed:
adjusted DI for TxStatusWaiter

Raunaque97 committed:
updating the indexer schema

Raunaque97 committed:
Merge remote-tracking branch 'origin/develop' into raun/retryMinaTransactions

Raunaque97 committed:
fixing errors

rpanic commented on packages/sequencer/src/protocol/baselayer/MinaBaseLayer.ts:
```diff
@@ -78,6 +94,17 @@ export class MinaBaseLayer
+        useClass: TxStatusWaiter,
+      },
+
+      L1TransactionDispatcherConfig: {
```
So - the pattern that we should use here is by using `useFactory` instead of `useValue` and then returning the Dispatcher instance - so inside the factory, resolve the dispatcher from the given dependencycontainer, then setting the config

rpanic commented on packages/sequencer/src/settlement/transactions/MinaTransactionSender.ts:
```diff
@@ -1,146 +1,125 @@
+    @inject("SettlementSigner") private readonly signer: MinaSigner,
+    @inject("FeeStrategy") private readonly feeStrategy: FeeStrategy,
+    private readonly dispatcher: L1TransactionDispatcher,
+    @inject("TxStatusWaiter") private readonly waiter: TxStatusWaiter
```
If I see it correctly, the waiter is only used inside the MinaTransactionSender and is not supposed to be overrideable - therefore I'd suggest resolving by class reference, not by token here

rpanic commented on packages/sequencer/src/settlement/transactions/MinaTransactionSender.ts:
```diff
@@ -1,146 +1,125 @@
+    private readonly dispatcher: L1TransactionDispatcher,
+    @inject("TxStatusWaiter") private readonly waiter: TxStatusWaiter
+  ) {
+    this.dispatcher.start();
```
We can use @startable() here (like closeable) and then start the dispatcher in there

rpanic commented on packages/sequencer/src/settlement/transactions/L1TransactionDispatcher.ts:
```diff
@@ -0,0 +1,279 @@
+        this.pollingTimeout.unref?.();
+      }
+    };
+    this.pollingTimeout = setTimeout(poll, this.config.pollIntervalMs);
```
I didn't know about unref - that's cool!
Why is this not a interval? Because of the unref or other reasons?

Raunaque97 replied:
if a poll() call takes more than  pollIntervalMs time, 2 calls can overlap

rpanic commented on packages/sequencer/src/settlement/transactions/L1TransactionRetryStrategy.ts:
```diff
@@ -0,0 +1,17 @@
+   * Optional retry delay. If provided, the dispatcher will schedule
+   * the next retry after the delay from the last sentAt time.
+   */
+  getRetryDelayMs?(record: PendingL1TransactionRecord): number;
```
What happens if the user doesn't specify that function? When will we retry?

Raunaque97 replied:
the retry will happen after `inclusionTimeoutMs` (default: 10 mins).
this delay is after checking for status ends.

rpanic commented on packages/sequencer/src/storage/repositories/PendingL1TransactionStorage.ts:
```diff
@@ -0,0 +1,48 @@
+  attempts: number;
+  status: PendingL1TransactionStatus;
+  transaction: Transaction<any, any>;
+  hash?: string;
```
https://github.com/o1-labs/o1js/blob/main/CHANGELOG.md#added
The PR that we requested from florian is now merged, we should use a nightly version of o1js for now until this is released but already now integrate it into this PR and use the hash everywhere

Raunaque97 committed:
Refactored DI for L1TransactionDispatcher, MinaTransactionSender

Raunaque97 committed:
using o1js nightly build,
removed hash field from PendingL1Transaction model

Raunaque97 committed:
updating tests for MinaTransactionSender

Raunaque97 committed:
prisma migrate indexer schema
